tools/challenge-parser/parser/plugins/utils/mdast-to-html.js

```diff
@@ -1,13 +1,71 @@
 const hastToHTML = require('hast-util-to-html');
 const { root } = require('mdast-builder');
 const mdastToHast = require('mdast-util-to-hast');
+const { parseChinesePattern } = require('./parse-chinese');
 
-function mdastToHTML(nodes) {
+/**
+ * Custom handler for Chinese inline code to render as ruby elements
+ * @param {object} state - The state object from mdast-util-to-hast
+ * @param {object} node - The inlineCode node
+ * @returns {object} Hast element node
+ */
+function chineseInlineCodeHandler(state, node) {
+  const parsed = parseChinesePattern(node.value);
+
+  if (parsed) {
+    return {
+      type: 'element',
+      tagName: 'ruby',
+      properties: {},
+      children: [
+        { type: 'text', value: parsed.hanzi },
+        {
+          type: 'element',
+          tagName: 'rp',
+          properties: {},
+          children: [{ type: 'text', value: '(' }]
+        },
+        {
+          type: 'element',
+          tagName: 'rt',
+          properties: {},
+          children: [{ type: 'text', value: parsed.pinyin }]
+        },
+        {
+          type: 'element',
+          tagName: 'rp',
+          properties: {},
+          children: [{ type: 'text', value: ')' }]
+        }
+      ]
+    };
+  }
+
+  return {
+    type: 'element',
+    // TODO: change this to span
+    // https://github.com/freeCodeCamp/language-curricula/issues/22
+    tagName: 'code',
+    properties: {},
+    children: [{ type: 'text', value: node.value }]
+  };
+}
+
+function mdastToHTML(nodes, options = {}) {
   if (!Array.isArray(nodes))
     throw Error('mdastToHTML expects an array argument');
   // - the 'nodes' are children, so first need embedding in a parent
 
-  return hastToHTML(mdastToHast(root(nodes), { allowDangerousHtml: true }), {
+  const { lang } = options;
```

Comment:
The only thing I'm unsure about is this. It seems a bit odd for a utility function to care about language.

I've spent a bit too long tinkering with this, but I think the way to go allow `mdastToHTML` to take options `function mdastToHTML(nodes, hastOptions = {}) {`, but pass the `handlers` directly. That way `mdastToHTML` remains a very simple utility.

In order to keep it easy to use, we can add a function `createMdastToHtml = lang => ...` that consumers can use if they want special language handling. Then, rather than passing around a `{ lang }` option, you can just use the custom stringifier.

If this all sounds somewhat reasonable, I've got a commit I can push so you can see what you think.

Reply:
And feel free to push your commit :D

Reply:
Given that it's only relevant to fill-in-the-blank, I think you can keep it as part of that plugin.  Even if there are multiple language specific  "handlers", it's totally fine to have them all in the same module.

It's very subjective, but if the module is coherent and only exposes a small api, I think it's very reasonable to keep the code together.

That's all a bit vague and high level, though, so I'm happy to discuss more once you've pushed some code.